File: contagem-malbarbo/contagem.py
#!/usr/bin/python
# encoding: utf-8

# for aninhado
def alg_a(palavras):
    saida = set()
    max_freq = 0
    for i in range(len(palavras)):
        freq = 0
        for j in range(len(palavras)):
            if palavras[i] == palavras[j]:
                freq += 1
            j += 1

        if freq == max_freq:
            saida.add(palavras[i])
        elif freq > max_freq:
            saida = set([palavras[i]])
            max_freq = freq

        i += 1
    return saida

# for aninhado onde j = i + 1
def alg_b(palavras):
    saida = set()
    max_freq = 0
    for i in range(len(palavras)):
        freq = 1
        for j in range(i + 1, len(palavras)):
            if palavras[i] == palavras[j]:
                freq += 1
            j += 1

        if freq == max_freq:
            saida.add(palavras[i])
        elif freq > max_freq:
            saida = set([palavras[i]])
            max_freq = freq

        i += 1
    return saida

# ordena e conta o tamanho da sequência com elementos repetidos
def alg_c(palavras):
    saida = set()
    palavras.sort()
    max_freq = 0
    i = 0
    while i < len(palavras):
        j = i + 1
        freq = 1
        while j < len(palavras) and palavras[i] == palavras[j]:
            freq += 1
            j += 1

        if freq == max_freq:
            saida.add(palavras[i])
        elif freq > max_freq:
            saida = set([palavras[i]])
            max_freq = freq

        i = j
    return saida

# usando tabela hash
def alg_d(palavras):
    saida = set()
    freq_hash = {}
    max_freq = 0
    for i in range(len(palavras)):
        if palavras[i] in freq_hash:
            freq_hash[palavras[i]] += 1
            freq = freq_hash[palavras[i]]
        else:
            freq_hash[palavras[i]] = 1
            freq = 1

        if freq == max_freq:
            saida.add(palavras[i])
        elif freq > max_freq:
            saida = set([palavras[i]])
            max_freq = freq

    return saida


def show_usage(argv):
    print 'Modo de usar: %s algorimo [arquivo]' % argv[0]
    print "    algoritmo é um dos valores:"
    print "        a - for aninhado "
    print "        b - for aninhado com j = i + 1"
    print "        c - ordenando a entrada"
    print "        d - tabela hash"
    print "    arquivo é o nome de um arquivo com uma palavra por linha"
    print "        se o nome do arquivo não for informado, a entrada"
    print "        padrão é utilizada"
        
if __name__ == '__main__':
    import sys
    if len(sys.argv) != 2 and len(sys.argv) != 3:
        show_usage(sys.argv)
        sys.exit(0)

    # parametro algoritmo
    alg_name = sys.argv[1]
    if alg_name == 'a':
        alg = alg_a
    elif alg_name == 'b':
        alg = alg_b
    elif alg_name == 'c':
        alg = alg_c
    elif alg_name == 'd':
        alg = alg_d
    else:
        print u'Nome do algoritmo inválido:', alg_name
        show_usage(sys.argv)
        sys.exit(0)

    # le a entrada
    filename = "-" if len(sys.argv) == 2 else sys.argv[2]
    if filename == "-":
        f = sys.stdin
    else:
        f = open(filename)
    palavras = [ s.strip() for s in f ]
    f.close()

    # executa o algoritmo
    import time
    start = time.clock()
    saida = alg(palavras)
    tempo = time.clock() - start

    # imprime a saída
    print tempo
    for p in saida:
        print p

File: contagem-malbarbo/contagem.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <list>
#include <unordered_map>
#include <string>
#include <ctime>
#include <cstdlib>

using namespace std;

struct Frequencia {
    string palavra;
    int freq;
};

list<Frequencia> alg_a(list<string> palavras) {
    list<Frequencia> freqs;
    int max_freq = 0;
    
    for (list<string>::iterator i = palavras.begin(); i != palavras.end(); i++) {
        int freq = 0;
        for(list<string>::iterator j = palavras.begin(); j != palavras.end(); j++) {
            if(*j == *i){
                freq++;
            }
        }
        
        if (freq == max_freq) {
            bool tem_freq = false;
            for (auto &freq_palavra : freqs) {
                if(freq_palavra.palavra == *i){
                    tem_freq = true;
                }
            }
            if (!tem_freq){
                Frequencia freq_palavra;
                freq_palavra.palavra = *i;
                freq_palavra.freq = freq;
                freqs.push_back(freq_palavra);
            }
        } else if (freq > max_freq) {
            max_freq = freq;
            freqs.clear();
            Frequencia freq_palavra;
            freq_palavra.palavra = *i;
            freq_palavra.freq = freq;
            freqs.push_back(freq_palavra);
        }
    }
    
    return freqs;
}

list<Frequencia> alg_b(list<string> palavras) {
    list<Frequencia> freqs;
    int max_freq = 0;
    
    for (list<string>::iterator i = palavras.begin(); i != palavras.end(); i++) {
        int freq = 0;
        for(list<string>::iterator j = i; j != palavras.end(); j++) {
            if(*j == *i){
                freq++;
            }
        }

        if (freq == max_freq) {
            Frequencia freq_palavra;
            freq_palavra.palavra = *i;
            freq_palavra.freq = freq;
            freqs.push_back(freq_palavra);
        } else if (freq > max_freq) {
            max_freq = freq;
            freqs.clear();
            Frequencia freq_palavra;
            freq_palavra.palavra = *i;
            freq_palavra.freq = freq;
            freqs.push_back(freq_palavra);
        }
    }
    
    return freqs;
}

list<Frequencia> alg_c(list<string> palavras) {
    list<Frequencia> freqs;
    int max_freq = 0;
    
    palavras.sort();
    
    list<string>::iterator i = palavras.begin();
    list<string>::iterator fim = palavras.end();
    
    while (i != fim) { 
        list<string>::iterator j = i;
        j++;
        
        int freq = 1;
        while (j != fim && *j == *i) {
            freq++;
            j++;
        }
        
        if (freq == max_freq) {
            Frequencia freq_palavra;
            freq_palavra.palavra = *i;
            freq_palavra.freq = freq;
            freqs.push_back(freq_palavra);
        } else if (freq > max_freq) {
            max_freq = freq;
            freqs.clear();
            Frequencia freq_palavra;
            freq_palavra.palavra = *i;
            freq_palavra.freq = freq;
            freqs.push_back(freq_palavra);
        }
        
        i = j;
    }
    
    return freqs;
}

list<Frequencia> alg_d(list<string> palavras) {
    list<Frequencia> freqs;
    unordered_map<string, int> freqs_hash;
    freqs_hash.reserve(palavras.size());
    int max_freq = 0;
    
    for (auto &palavra : palavras) {
        freqs_hash[palavra]++;
        int freq = freqs_hash[palavra];
        
        if (freq == max_freq) {
            Frequencia freq_palavra;
            freq_palavra.palavra = palavra;
            freq_palavra.freq = freq;
            freqs.push_back(freq_palavra);
        } else if (freq > max_freq) {
            max_freq = freq;
            freqs.clear();
            Frequencia freq_palavra;
            freq_palavra.palavra = palavra;
            freq_palavra.freq = freq;
            freqs.push_back(freq_palavra);
        }
    }
    
    return freqs;
}

void print(list<string> palavras) {
    cout << "(palavra):" << "\n";
    for (auto &palavra : palavras) {
        cout << "(" << palavra << ")" << "\n";
    }
}

void print(list<Frequencia> freqs) {
    cout << "(palavra, frequencia):" << "\n";
    for (auto &freq : freqs) {
        cout << "(" << freq.palavra << "," << freq.freq << ")" << "\n";
    }
}

void print_palavras(list<Frequencia> freqs){
    for(auto &freq : freqs){
        cout << freq.palavra << endl;
    }
}

void read_text_file(const char *filename, list<string>& v) {
    ifstream fin(filename);
    istream *in;
    if (string("-") == filename) {
        in = &cin;
    } else {
        in = &fin;
    }
    if (in->fail()) {
        cout << "Não foi possivél abrir o arquivo: " << filename << endl;
        exit(0);
    }
    string s;
    while (getline(*in, s)) {
        v.push_back(s);
    }
}

int str2int(string s) {
    istringstream buffer(s);
    int i;
    buffer >> i;
    return i;
}

void show_usage(const char* argv[]) {
    cout << "Modo de usar: " << argv[0] << " algoritmo [arquivo]" << endl;
    cout << "    algoritmo é um dos valores:" << endl;
    cout << "        a - for aninhado " <<  endl;
    cout << "        b - for aninhado com j = i + 1" << endl;
    cout << "        c - ordenando a entrada" << endl;
    cout << "        d - tabela hash" << endl;
    cout << "    arquivo é o nome de um arquivo com uma palavra por linha" << endl;
    cout << "        se o nome do arquivo não for informado, a entrada" << endl;
    cout << "        padrão é utilizada" << endl;
}

int main(int argc, const char* argv[]) {
    if (argc != 2 && argc != 3) {
        show_usage(argv);
        return 0;
    }

    // parametro algoritmo
    string alg_name = argv[1];
    list<Frequencia> (*alg)(list<string>);

    if (alg_name == "a") {
        alg = alg_a;
    } else if (alg_name == "b") {
        alg = alg_b;
    } else if (alg_name == "c") {
        alg = alg_c;
    } else if (alg_name == "d") {
        alg = alg_d;
    } else {
        cout << "Nome do algoritmo inválido: " << alg_name << endl;
        show_usage(argv);
        return 1;
    }
        
    // parametro arquivo
    const char *filename = argc == 2 ? "-" : argv[2];
    list<string> palavras;
    read_text_file(filename, palavras);

    // executa o algoritmo
    clock_t start = clock();
    list<Frequencia> saida = alg(palavras);
    double tempo = (double)(clock() - start) / CLOCKS_PER_SEC;
    cout << tempo << endl << flush;

    // imprime a saida
    print_palavras(saida);
    return 0;
}
